File: fit_ellipse.cpp


/*
input:
coordinate: the coordinate of edge points on pupil circle.
it format is
[x_1, y_1;
 x_2, y_2;
   .,   .;
   .,   .;   
 x_n, y_n;
   .,   .;
   .,   .;   
 x_N, y_N;
];
*/
#include <math.h>
#include "fit_ellipse.h"
#include "matrix.h"

matrix fit_ellipse(const matrix& coord)
{
    matrix D1=(coord.columnVector(0)*coord.columnVector(0)).appendright(coord.columnVector(0)*coord.columnVector(1)).appendright(coord.columnVector(1)*coord.columnVector(1));
    matrix D2=coord.columnVector(0).appendright(coord.columnVector(1)).appendright(matrix::ones(coord.height(),1));
    matrix S1=matrix::crossProduct(D1.transform(), D1);
    matrix S2=matrix::crossProduct(D1.transform(), D2);
    matrix S3=matrix::crossProduct(D2.transform(), D2);
    matrix T=matrix::crossDivision(S3, S2.transform())*(-1);
    matrix M=S1+matrix::crossProduct(S2, T);
    matrix tmp(M.height(), M.width());
    for(unsigned i=0; i<M.height(); i++)
    {
    	for(unsigned j=0; j<M.width(); j++)
    	{
            if(i==0)
                tmp.setvalue(i,j, 0.5*M.getvalue(2,j));
            else if(i==1)
                tmp.setvalue(i,j, -M.getvalue(1,j));
            else
                tmp.setvalue(i,j, 0.5*M.getvalue(0,j));
    	}
    }

    matrix eigv=tmp.eigenvector();
    double cond=0;
    matrix a1;
    for(unsigned j=0; j<M.width(); j++)
    {         
           cond=4*eigv.getvalue(0, j)*eigv.getvalue(2, j)-eigv.getvalue(1, j)*eigv.getvalue(1, j);
    	   if(cond>0)
    	    a1=eigv.columnVector(j);
    }
    matrix under=matrix::crossDivision(T, a1);
    double a=a1.getvalue(0,0);
    double b=a1.getvalue(0,1)/2;
    double c=a1.getvalue(0,2);
    double d=under.getvalue(0,0)/2;
    double f=under.getvalue(0,1)/2;
    double g=under.getvalue(0,2);

    //ellipse center
    double num=b*b-a*c;
    double XCenter=(c*d-b*f)/num;
    double YCenter=(a*f-b*d)/num;

    //ellipse angle of rotation.
    double theta=0.5*atan(2*b/(a-c));

    //ellipse axis length
    double up=2*(a*f*f+c*d*d+g*b*b-2*b*d*f-a*c*g);
    double down1=(b*b-a*c)*((c-a)*sqrt(1+4*b*b/((a-c)*(a-c)))-(c+a));
    double down2=(b*b-a*c)*((a-c)*sqrt(1+4*b*b/((a-c)*(a-c)))-(c+a));
    double xRadius=sqrt(up/down1);
    double yRadius=sqrt(up/down2);

    matrix parameter(1,5);
    parameter.setvalue(0, 0, XCenter);
    parameter.setvalue(0, 1, YCenter);
    parameter.setvalue(0, 2, theta);
    parameter.setvalue(0, 3, xRadius);
    parameter.setvalue(0, 4, yRadius);
    return parameter;
}

File: image/sample1/fit_ellipse.m
function [xCenter, yCenter, xRadius, yRadius, theta]=fit_ellipse(x, y)

%     Direct ellipse fitting, proposed in article
%     Hal?r R, Flusser J. "Numerically stable direct least squares fitting of ellipses"[C]
%     Proc. 6th International Conference in Central Europe on Computer Graphics and Visualization. 
%     WSCG. 1998, 98: 125-132.
%
%     Our code is based on a numerically stable version
%     of this fit published by R. Halir and J. Flusser
%
%     Input: These two values come from the output of LineRemove function.
%     x: coordinate value along x-axis
%     y: coordinate value along y-axis
%
%     Output:
%     xCenter: the center coordinate value along x-axis
%     yCenter: the center coordinate value along y-axis
%     xRadius: the radius value along x-axis
%     xRadius: the radius value along y-axis

%     This is a fast non-iterative ellipse fit.  It returns ellipses only, even if points are
%     better approximated by a hyperbola. It is somewhat biased toward smaller ellipses.




D1=[x.^2, x.*y, y.^2];
D2=[x, y, ones(size(x))];
S1=D1'*D1;
S2=D1'*D2;
S3=D2'*D2;
T=-inv(S3)*S2';
M=S1+S2*T;
M=[M(3,:)./2; -M(2, :); M(1, :)./2];
[evec, ~]=eig(M);
cond=4*evec(1, :).*evec(3, :)-evec(2, :).^2;
a1=evec(:, find(cond>0));
A=[a1;T*a1];

a=A(1);
b=A(2)/2;
c=A(3);
d=A(4)/2;
f=A(5)/2;
g=A(6);
%% ellipse Center
num=b^2-a*c;
xCenter=(c*d-b*f)/num;
yCenter=(a*f-b*d)/num;

%% ellipse angle of rotation
theta=0.5*atan(2*b/(a-c));


%% ellipse axis length
up = 2*(a*f*f+c*d*d+g*b*b-2*b*d*f-a*c*g);
down1=(b*b-a*c)*((c-a)*sqrt(1+4*b*b/((a-c)*(a-c)))-(c+a));
down2=(b*b-a*c)*((a-c)*sqrt(1+4*b*b/((a-c)*(a-c)))-(c+a));
xRadius=sqrt(up/down1);
yRadius=sqrt(up/down2);

end

